packages/teleterm/src/ui/DocumentCluster/ClusterResources/Databases/Databases.tsx

```diff
@@ -60,13 +65,26 @@ function DatabaseList(props: State) {
   );
 }
 
-function renderConnectButton(onConnect: (user: string) => void) {
+function ConnectButton({
+  dbUri,
+  onConnect,
+}: {
+  dbUri: string;
+  onConnect: (user: string) => void;
+}) {
+  const { getUsersAttempt, getUsers } = useDatabaseUsers(dbUri);
+
   return (
     <Cell align="right">
       <MenuLoginTheme>
-        <MenuLogin
+        <MenuLoginAsync
           placeholder="Enter username…"
-          getLoginItems={() => []}
+          getLoginItemsAttempt={getUsersAttempt}
+          getLoginItems={() => {
```

Comment:
I feel like we can simplify it. I see no benefit from passing an attempt to the component (and it increases usage complexity). Why the component can't handle data fetching on its own? 
Current solution triggers the request only once, but we can achieve the same effect by simply wrapping a promise into `memoize` here.

Reply:
> Why the component can't handle data fetching on its own?

My instinct is to push any async stuff to the outermost layer of the app, as far as possible. I was also thinking about a scenario where we'd want to share async data between `MenuLoginAsync` and some other compontent. But even in that case, it should be possible to just pass a promise to `MenuLogin` that was prepared somewhere else. 🤔

And I agree that the usage complexity would be lower then. Because then essentially I'd do

```jsx
<MenuLogin
  getLoginItems={someFunctionReturningPromise}
/>
```

right?

> Current solution triggers the request only once, but we can achieve the same effect by simply wrapping a promise into `memoize` here.

We probably wouldn't even need memoize, if `useAsync` was called from inside `MenuLogin` then we could check the status of the attempt object just as I do know, I think.

I'd need to make to make a function component out of `MenuLogin` to use `useAsync` in it, wouldn't I?

Reply:
I agree, if there would be a case to share some data, then we could create a special promise for that. 

> Because then essentially I'd do
`<MenuLogin
  getLoginItems={someFunctionReturningPromise}
/>`
right?

Yes

> We probably wouldn't even need memoize, if useAsync was called from inside MenuLogin then we could check the status of the attempt object just as I do know, I think.

Ah right, I forgot that `MenuLogin` isn't destroyed when is closed.

> I'd need to make to make a function component out of MenuLogin to use useAsync in it, wouldn't I?

Yes, you can't use hooks in class components.